numpyro/contrib/autoguide/__init__.py

```diff
@@ -0,0 +1,256 @@
+        sites in the model.
+
+        :param jax.random.PRNGKey rng: PRNG seed.
+        :param opt_state: Current state of the optmizer.
+        :param sample_shape: (keyword argument) shape of samples to be drawn.
+        :return: batch of samples from the approximate posterior.
+        """
+        raise NotImplementedError
+
+    def _sample_latent(self, *args, **kwargs):
+        """
+        Samples an encoded latent given the same ``*args, **kwargs`` as the
+        base ``model``.
+        """
+        raise NotImplementedError
+
+    def _setup_prototype(self, *args, **kwargs):
+        # run the model so we can inspect its structure
+        self.prototype_trace = block(trace(self.model).get_trace)(*args, **kwargs)
+
+
+class AutoContinuous(AutoGuide):
+    """
+    Base class for implementations of continuous-valued Automatic
+    Differentiation Variational Inference [1].
+
+    Each derived class implements its own :meth:`get_posterior` method.
+
+    Assumes model structure and latent dimension are fixed, and all latent
+    variables are continuous.
+
+    :param callable model: a Pyro model
+
+    Reference:
+
+    [1] `Automatic Differentiation Variational Inference`,
+        Alp Kucukelbir, Dustin Tran, Rajesh Ranganath, Andrew Gelman, David M.
+        Blei
+
+    :param callable model: A Pyro model.
+    :param callable init_loc_fn: A per-site initialization function.
+        See :ref:`autoguide-initialization` section for available functions.
+    """
+    def __init__(self, rng, model, get_params_fn, prefix="auto", init_loc_fn=init_to_median):
+        model = substitute(model, substitute_fn=block(seed(init_loc_fn, rng)))
+        super(AutoContinuous, self).__init__(model, get_params_fn, prefix=prefix)
+
+    def _setup_prototype(self, *args, **kwargs):
+        super(AutoContinuous, self)._setup_prototype(*args, **kwargs)
+        self._unconstrained_values = {}
+        self._inv_transforms = {}
+        for name, site in self.prototype_trace.items():
+            if site['type'] == 'sample' and not site['is_observed']:
+                # Collect the shapes of unconstrained values.
+                # These may differ from the shapes of constrained values.
+                transform = biject_to(site['fn'].support)
+                unconstrained_val = transform.inv(site['value'])
+                self._inv_transforms[name] = transform
+                self._unconstrained_values[name] = unconstrained_val
+
+        self.latent_size = sum(np.size(x) for x in self._unconstrained_values.values())
+        if self.latent_size == 0:
+            raise RuntimeError('{} found no latent variables; Use an empty guide instead'.format(type(self).__name__))
+        _, self.unravel_fn = ravel_pytree(self._unconstrained_values)
+
+    def sample_posterior(self, rng, opt_state, *args, **kwargs):
+        raise NotImplementedError
+
+    def _sample_latent(self, *args, **kwargs):
+        raise NotImplementedError
```

Comment:
nit: we can remove this because it is handled in AutoGuide

Reply:
Done. The reason for having this was to indicate that AutoContinuous is meant to be subclasses, but moved to using Abstract Base Class for this instead.